File: bindings/typescript/examples/jsonGrammar.ts
/**
 * Compile the built-in JSON grammar and accept a sample JSON string.
 *
 * Build the NAPI addon first (`cargo tools build typescript --targets host`),
 * then run: `pnpm exec tsx examples/jsonGrammar.ts`
 */
import {
  TokenizerInfo,
  GrammarCompiler,
  GrammarMatcher,
} from "../src/index";

const encodedVocab: string[] = [];
for (let i = 0; i < 256; i++) {
  encodedVocab.push(String.fromCharCode(i));
}
encodedVocab.push("</s>");

const tokenizerInfo = new TokenizerInfo(
  encodedVocab,
  1, // VocabType.ByteFallback
  null,
  [256],
  false
);
const compiler = new GrammarCompiler(tokenizerInfo, 8, true, -1);
const compiled = compiler.compileBuiltinJsonGrammar();
const matcher = new GrammarMatcher(compiled, null, false, -1);

const sample = '{"name":"Ada","role":"student"}';
if (!matcher.acceptString(sample, false)) {
  throw new Error(`grammar rejected: ${sample}`);
}
console.log(sample);
console.log(
  `completed=${matcher.isCompleted()} terminated=${matcher.isTerminated()}`
);

File: bindings/python/examples/json_grammar.py
"""Compile the built-in JSON grammar and accept a sample JSON string."""

import xgrammar as xgr

# Minimal byte-level vocab: one token per byte 0..=255, plus a stop token.
encoded_vocab = [bytes([i]).decode("latin1") for i in range(256)] + ["</s>"]
tokenizer_info = xgr.TokenizerInfo(
    encoded_vocab,
    vocab_type=xgr.VocabType.BYTE_FALLBACK,
    stop_token_ids=[256],
)
compiler = xgr.GrammarCompiler(tokenizer_info)
compiled = compiler.compile_builtin_json_grammar()
matcher = xgr.GrammarMatcher(compiled)

sample = '{"name":"Ada","role":"student"}'
assert matcher.accept_string(sample), f"grammar rejected: {sample}"
print(sample)
print(f"completed={matcher.is_completed()} terminated={matcher.is_terminated()}")
